### PR TITLE
Move RWLock alloc to use NdisHandle

### PR DESCRIPTION
**Pull request by jkunkee**
This is required for Windows 7 per the documentation for `NdisAllocateRWLock`.

### CODE REVIEW

File: src/tapdrvr.c [hunk 1]
```diff
@@ -169,6 +159,18 @@ Routine Description:
         if (NDIS_STATUS_SUCCESS == status)
         {
             DEBUGP (("[TAP] Registered miniport successfully\n"));
+
+            //
+            // This lock protects the AdapterList.
+            //
+            GlobalData.Lock = NdisAllocateRWLock(GlobalData.NdisDriverHandle);
+            if (GlobalData.Lock == NULL)
```
Comment (selvanair) [inline src/tapdrvr.c]: This doesn't look right. I think its be too late to create the lock here. NdisMRegisterMiniportDriver() will call AdapterCreate() callback (the InitializeHandlerEx callback) which will call tapAdapterContextAddToGlobalList() which needs the lock.

Comment (jkunkee) [reply src/tapdrvr.c]: Hang on...so NdisAllocateRWLock requires the handle returned by NdisMRegisterMiniportDriver, and NdisMRegisterMiniportDriver expects NdisAllocateRWLock to have been called already? That doesn't seem right, but I've seen worse...

Are you sure NdisMRegisterMiniportDriver calls AdapterCreate? I thought adapter creation was after DriverEntry returned. I can step through this code path tomorrow.

Comment (selvanair) [reply src/tapdrvr.c]: See the docs on NdisMRegisterMiniportDriver for which callbacks are called from it. InitializeHandlerEx() will definitely get called and that is AdapterCreate() for us.

A way out is to implement SetOptions and create the lock there. May be that's how its supposed to be used in 6.20 when NULL handle is not an option (Win7).

**Comment by cron2** [comment]
Hi,

On Wed, May 30, 2018 at 06:28:42PM -0700, Selva Nair wrote:
> NdisMRegisterMiniportDriver() will call AdapterCreate() callback 
> (the InitializeHandlerEx callback) which will call 
> tapAdapterContextAddToGlobalList() which needs the lock. 

Can we create the lock in AdapterCreate()?

(Well spotted, btw, I thought this would be the easy and logical way to
adapt things...)

gert

-- 
"If was one thing all people took for granted, was conviction that if you 
 feed honest figures into a computer, honest figures come out. Never doubted 
 it myself till I met a computer with a sense of humor."
                             Robert A. Heinlein, The Moon is a Harsh Mistress

Gert Doering - Munich, Germany                             gert@greenie.muc.de

**Comment by selvanair** [comment]
My remark is based on the following in the docs for NdisMRegisterMiniportDriver() here: https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/content/ndis/nf-ndis-ndismregisterminiportdriver

_After a driver calls NdisMRegisterMiniportDriver, the driver should be prepared to be called back at the MiniportInitializeEx function that is specified in the MiniportDriverCharacteristics parameter._

But another source (https://docs.microsoft.com/en-us/windows-hardware/drivers/network/initializing-a-miniport-driver) has this:

_Drivers that call NdisMRegisterMiniportDriver must be prepared for NDIS to call their MiniportInitializeEx functions **any time after** DriverEntry **returns**._  (emphasis mine).

If the latter is the right description we are safe. So if @jkunkee can confirm that call never happens before Driver entry returns, we are ok as is.

**Comment by jkunkee** [comment]
[This sample](https://github.com/Microsoft/Windows-driver-samples/blob/master/network/ndis/netvmini/6x/miniport.c) has lock initialization after the call to NdisMRegisterMiniportDriver, during DriverEntry, so that's good. I can still ask the NDIS team for clarification of the docs.

**Comment by selvanair** [comment]
Yeah, I've seen that sample. But the docs seem to imply otherwise --- it would be safe to get some clarification. Thanks.

**Comment by jkunkee** [comment]
Per the NDIS team's response, this is the correct version:
<br/>
_Drivers that call NdisMRegisterMiniportDriver must be prepared for NDIS to call their MiniportInitializeEx functions any time after DriverEntry returns._

**Comment by selvanair** [comment]
That settles it, I guess.

One more thing: if the NdisAllocateRWLock fails, NdisFreeRWLock() will get called with A NULL (in unload).  This may be ok (?), but better check for NULL and also set the lock pointer to NULL after freeing.

Looks good otherwise.

**Comment by jkunkee** [comment]
Done and done.

**Comment by mattock** [comment]
@jkunkee would it be possible to harmonize the two documents above? Or are those outside the jurisdiction of the NDIS team? Just thinking of the future poor souls who get confused about this :smile:.

**Comment by cron2** [comment]
I take Selva's last comment as an ACK-if, and the change has been made :-) - and my understanding of code and clarified documentation also says "this is good, it should go in".  So, done.

**Pull request merged**